### PR TITLE
[Reopen] Support frondend in E2E transformer

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
Sorry, i mistakenly deleted branch. Original PR is #869
I modified E2E transformer to support frontend.

- [x] Add tests
- [x] Implement FrondendInterface and support dynamic importing scheme for fronend.

### CODE REVIEW

**Comment by kamo-naoyuki** [comment]
I changed FrontendASRInterface to generalize featdim.

```python
frontend_asr = FrontendASR(infeatdim, args)
asr_model = E2E(frontend.featdim, args)
frontend_asr.register_asr(asr_model)
```

I changed the options to reflect the @ShigekiKarita 's comment. e.g. `--blayers` -> `--beamformer-layers`

**Comment by kamo-naoyuki** [comment]
I also deleted all docstring of FrontendASR, then it can automatically inherits the base class.

**Comment by kamo-naoyuki** [comment]
I appended egs/an4/asr2 for testing frontend system. 
(I'm also planning to unify multichannel recipe to asr2, so I named it as asr2.)

It's not related to this PR, this recipe achieved better results (CER=4.1 WER=7.9)

**Comment by kamo-naoyuki** [comment]
I moved the arguments of `asr_model` and `mt_model` for E2E rnn class to `easpnet/asr/pytorch_backend/asr.py` because it doesn't satisfy the interface constrain. Current v0.5.0 version doesn't work even E2E transformer.

File: espnet/asr/pytorch_backend/asr.py [hunk 1]
```diff
@@ -302,28 +275,46 @@ def train(args):
...
+    if args.frontend_module is not None:
+        # The concrete class for Fronend is here: espnet.nets.pytorch_backend.frontend_asr
+        frontend_class = dynamic_import(args.frontend_module)
+        assert issubclass(frontend_class, FrontendASRInterface), frontend_class
+        # Wrap model by frontend_class.
+        frontend_asr = frontend_class(idim, args)
+        _idim = frontend_asr.featdim
+    else:
+        frontend_asr = None
+        _idim = idim
```
Comment (kamo-naoyuki) [inline espnet/asr/pytorch_backend/asr.py]: Maybe this part looks complicated, but I think it's better to refactor it with the other part as a whole.
`asr.py` has too many bad coding, for example, the accessing  to the not interface attribute.

```
reporter = model.reporter
```

However, many of them come from chainer trainer restriction, so @ShigekiKarita 's Trainer is desired to be merged quickly.

**Comment by kamo-naoyuki** [comment]
I remembered pytorch.DataParallel doesn't support namedtuple yet. I'll restore it to dict.

File: espnet/nets/pytorch_backend/e2e_asr.py [hunk 2]
```diff
@@ -101,19 +96,6 @@ def __init__(self, idim, odim, args, asr_model=None, mt_model=None):
         # speech translation related
         self.replace_sos = args.replace_sos
 
-        if args.use_frontend:
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_asr.py]: I actually could not fully track where the front end processing is performed. Could you point out it to me?
I prefer to have simple sequential processing in the main training loop.

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/e2e_asr.py]: > I actually could not fully track where the front end processing is performed. Could you point out it to me?

I changed my mind. 
I put back to frontend class in E2E and changed the interface slightly. Maybe you can get the procedure now.

```python
frontend = FrontendASR(idim, args)
asr = E2E(frontend.featdim, odim, args)
asr.register_frontend(frontend)
```

**Comment by kamo-naoyuki** [comment]
@sw005320 This PR was done. I'll send the chime4 result and the config file after this PR.

File: egs/an4/asr2/cmd.sh [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    export train_cmd="ssh.pl"
+    export cuda_cmd="ssh.pl"
+    export decode_cmd="ssh.pl"
+
+# This is an example of specifying several unique options in the JHU CLSP cluster setup.
+# Users can modify/add their own command options according to their cluster environments.
+elif [ "${cmd_backend}" = jhu ]; then
+
+    export train_cmd="queue.pl --mem 2G"
+    export cuda_cmd="queue.pl --mem 2G --gpu 1 --config conf/gpu.conf"
```
Comment (sw005320) [inline egs/an4/asr2/cmd.sh]:
```suggestion
    export cuda_cmd="queue-freegpu.pl --mem 2G --gpu 1 --config conf/gpu.conf"
```

File: egs/an4/asr2/run.sh [hunk 4]
```diff
@@ -0,0 +1,246 @@
+#!/bin/bash
+
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+. ./path.sh
```
Comment (sw005320) [inline egs/an4/asr2/run.sh]:
```suggestion
. ./path.sh || exit 1;
```

File: egs/an4/asr2/run.sh [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+. ./cmd.sh
```
Comment (sw005320) [inline egs/an4/asr2/run.sh]:
```suggestion
. ./cmd.sh || exit 1;
```

File: egs/an4/asr2/run.sh [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+datadir=./downloads
+an4_root=${datadir}/an4
+data_url=http://www.speech.cs.cmu.edu/databases/an4/
+
+# exp tag
+tag="" # tag for managing experiments.
+
+. utils/parse_options.sh || exit 1;
+
+. ./path.sh
```
Comment (sw005320) [inline egs/an4/asr2/run.sh]: Remove lines 43 and 44. They're duplicated.

File: espnet/nets/pytorch_backend/e2e_asr.py [hunk 7]
```diff
@@ -126,24 +108,6 @@ def __init__(self, idim, odim, args, asr_model=None, mt_model=None):
         # weight initialization
         self.init_like_chainer()
 
-        # pre-training w/ ASR encoder and NMT decoder
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_asr.py]: I think it should not be removed. This is used by speech translation

**Comment by sw005320** [comment]
@Xiaofei-Wang, could you also review this PR? It would affect your REVERB multichannel recipe.

**Comment by Xiaofei-Wang** [comment]
> @Xiaofei-Wang, could you also review this PR? It would affect your REVERB multichannel recipe.

Sure.

File: egs/an4/asr2/conf/tuning/train_mtlalpha0.5.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# network architecture
+frontend-module: "espnet.nets.pytorch_backend.frontend_asr:FrontendASR"
```
Comment (Xiaofei-Wang) [inline egs/an4/asr2/conf/tuning/train_mtlalpha0.5.yaml]: As this recipe is placed for testing of frontend recipe, WPE should also be tested. What's your opinion?

**Comment by sw005320** [comment]
@kamo-naoyuki, any progress? Also, I changed this PR target from the baster to v.0.6.0 branch since it includes several refactoring.

**Pull request closed without merging**